subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -58,6 +59,24 @@ public <K, V> CrossBuildInMemoryCache<K, V> newCache() {
         return cache;
     }
 
+    @Override
+    public <K, V> CrossBuildInMemoryCache<K, V> newCache(Consumer<V> onReuse) {
+        DefaultCrossBuildInMemoryCache<K, V> cache = new DefaultCrossBuildInMemoryCache<K, V>(new HashMap<>()) {
+            @Nullable
+            @Override
+            protected V maybeGetRetainedValue(K key) {
+                V v = super.maybeGetRetainedValue(key);
+                if (v != null) {
+                    onReuse.accept(v);
```

Comment:
❓ Is it too bad that this callback happens under the cache lock?

Reply:
I assume it depends on what the callback is doing. It the current use case, class loader cache, it feels okay. I'd say that we should at least explicitly document the contract so future readers don't have to figure this out by themselves.

Reply:
> I'd say that we should at least explicitly document the contract so future readers don't have to figure this out by themselves.

Right! Done in be0c72bcbb624d23279a3024425e4277380348a4.